src/DIRAC/Resources/IdProvider/IdProvider.py

```diff
@@ -26,3 +33,27 @@ def setParameters(self, parameters):
         """
         self.parameters = parameters
         self.name = parameters.get("ProviderName")
+
+    def getGroupScopes(self, group: str) -> list:
+        """Get group scopes
+
+        :param group: DIRAC group
+        """
+        idPScope = getGroupOption(group, "IdPRole")
+        return scope_to_list(idPScope) if idPScope else []
+
+    def getScopeGroups(self, scope: str) -> list:
+        """Get DIRAC groups related to scope"""
+        groups = []
+        for group in getAllGroups():
+            if (g_scope := self.getGroupScopes(group)) and set(g_scope).issubset(scope_to_list(scope)):
+                groups.append(group)
+        return groups
+
+    def refreshToken(self, refreshToken=None, **kwargs):
+
+        return S_ERROR("Not implemented")
```

Comment:
How about making this an abstract base class so it can only be used if you have an actual implementation?

```
In [1]: from abc import ABCMeta, abstractmethod
   ...:
   ...: class AbstractIdProvider(metaclass=ABCMeta):
   ...:     @abstractmethod
   ...:     def refreshToken(self, refreshToken=None, **kwargs):
   ...:         ...
   ...:
   ...: class CheckInIdProvider(AbstractIdProvider):
   ...:     def refreshToken(self, refreshToken=None, **kwargs):
   ...:         print("Hello world")
   ...:

In [2]: AbstractIdProvider()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 AbstractIdProvider()

TypeError: Can't instantiate abstract class AbstractIdProvider with abstract method refreshToken

In [3]: CheckInIdProvider()
Out[3]: <__main__.CheckInIdProvider at 0x7f1293768a90>
```

Reply:
What is a real advantage of creating an abstract class ?

Reply:
* When we configure it incorrectly we don't get `{"OK": False, "Message": "Not implemented"}` in the logs with no idea where it came from
* It becomes impossible to instantiate the base class so you get an error as early as possible
* Static analysers are more likely to be able to understand what you're doing
* The abstract methods don't need to return failures so you can still use `super()` to call them and get correct MRO (though I hope we don't have any class hierarchies that are complex enough for that to matter)

Reply:
MRO ?

Reply:
Updated